### PR TITLE
Reverse Proxy Support

### PR DESCRIPTION
**Pull request by MatthijsKrempel**
Module for configuring OrchardCore hosting behind a reverse proxy.

Next versions will be able to manage networks and known proxies for more security.

### CODE REVIEW

**Comment by sebastienros** [comment]
Can you explain what it means to configure Orchard when it's behind a proxy? How should it be different? The hidden question is why don't these settings also apply in a normal case?

**Comment by MatthijsKrempel** [comment]
@sebastienros We currently use a single proxy to manage SSL for multiple sites under a public domain. We can run each instance of a public website on HTTP locally, this removes the burden of managing SSL certificates on the various platforms. OrchardCore needs to be aware that the reverse proxy handles the SSL, so we enable the X-Froward- headers middelware. I can imagine this is also helpful when you want to cloud host OrchardCore solutions behind cloud traffic managers.

**Comment by awyl** [comment]
The asp http overrides middleware only look at X forwarded headers if proxy network or proxy address is set in its configuration.

**Comment by MatthijsKrempel** [comment]
@awyl You can clear those options so it allows anything, I intend to implement these features later.

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Services/IReverseProxyService.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using System.Threading.Tasks;
+using OrchardCore.ReverseProxy.Settings;
+
+namespace OrchardCore.ReverseProxy.Services
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ReverseProxy/Services/IReverseProxyService.cs]: Theorically, as long as the interface is kept in the module is kept in the module, you don't need an interface.

File: src/OrchardCore.Modules/OrchardCore.ReverseProxy/Views/ReverseProxySettings.Edit.cshtml [hunk 2]
```diff
@@ -0,0 +1,34 @@
+@model ReverseProxySettingsViewModel
+
+<fieldset class="form-group" asp-validation-class-for="EnableXForwardedFor">
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ReverseProxy/Views/ReverseProxySettings.Edit.cshtml]: Maybe add a link to the reference documentation https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.1

**Comment by awyl** [comment]
Strange.  My other websites, which are plain aspnet core, I need to set the proxy options to get real client ip.  Host header, no problem though.  I checked source code last time.

File: src/OrchardCore/OrchardCore.Application.Cms.Targets/OrchardCore.Application.Cms.Targets.csproj [hunk 3]
```diff
@@ -12,6 +12,7 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\..\OrchardCore.Modules\OrchardCore.ReverseProxy\OrchardCore.ReverseProxy.csproj" />
```
Comment (sebastienros) [inline src/OrchardCore/OrchardCore.Application.Cms.Targets/OrchardCore.Application.Cms.Targets.csproj]: It's a module, add it in the module sections, ordered alphabetically

**Comment by sebastienros** [comment]
Was thinking about it, and now I am wondering if it should actually be a module. In the case of multi-tenancy, wouldn't the settings be the same for all tenants? However, does the middleware instance have to be specific to each tenant pipeline, or could it be a common one before the tenants ones?

**Comment by MatthijsKrempel** [comment]
Agree, I highly doubt that settings would differ from tenant to tenant.
I copied the approach from the Https module, which also has this same issue.
I think we are looking at a third module type; infrastructure (applicable to all tenants).
These should only be configurable from the main tenant.
What I can do for now, is look if the configuration is present in the appsettings.json and activate the module if this is the case.

**Comment by sebastienros** [comment]
We can resolve the settings and only do something based on the default name, this is what we do for Tenants management. We could also add some feature to add a specific attribute to dependencies.

**Comment by sebastienros** [comment]
And the feature could register something at the main host level (at least I assume it's possible, not sure). If not possible then it can't be a module, it would have to be defined by code, in the main app's Startup.

**Comment by MatthijsKrempel** [comment]
The problem I have with defining it in the app's Startup is that development and production code would differ. Would be in favor of appsettings.json or configure it on the default tenant only.

When I started I assumed this was the way to go, since the Https module was already in place?

**Comment by sebastienros** [comment]
appsettings is fine, you still need to write the code in startup.

**Comment by MatthijsKrempel** [comment]
I am currently also working on a CORS module (we need it for Swagger via OIDC), will have a look if I can also apply the requirement for managing CORS settings from the default tenant only.
Code in the startup defeats the purpose for us to use OrchardCore. We really like the modularity and writing code in the startup negates that point, so we really would like to see it being a feature that is turned on and configured.

**Comment by sebastienros** [comment]
Ok, then let's keep it as a module, and people can still change the startup if they need to implement it globally.

**Pull request merged**